src/lib/cert/x509/name_constraint.cpp

```diff
@@ -0,0 +1,269 @@
+/*
+* X.509 Name Constraint
+* (C) 2015 Kai Michaelis
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#include <botan/name_constraint.h>
+#include <botan/ber_dec.h>
+#include <botan/der_enc.h>
+#include <botan/charset.h>
+#include <botan/loadstor.h>
+#include <botan/x509_dn.h>
+#include <botan/x509cert.h>
+#include <sstream>
+
+namespace Botan {
+
+GeneralName::GeneralName(const std::string& v) : GeneralName()
+   {
+   size_t p = v.find(':');
+
+   if(p != std::string::npos)
+      {
+      m_type = v.substr(0,p);
+      m_name = v.substr(p + 1,std::string::npos);
+      }
+   else
+      {
+      throw Invalid_Argument("Failed to decode Name Constraint");
+      }
+   }
+
+void GeneralName::encode_into(class DER_Encoder&) const
+   {
+   throw std::runtime_error("General Name encoding not implemented");
+   }
+
+void GeneralName::decode_from(class BER_Decoder& ber)
+   {
+   BER_Object obj = ber.get_next_object();
+   if((obj.class_tag != CONTEXT_SPECIFIC) &&
+      (obj.class_tag != (CONTEXT_SPECIFIC | CONSTRUCTED)))
+      throw Decoding_Error("Invalid class tag while decoding GeneralName");
+
+   const ASN1_Tag tag = obj.type_tag;
+
+   if(tag == 1 || tag == 2 || tag == 6)
+      {
+      m_name = Charset::transcode(ASN1::to_string(obj),LATIN1_CHARSET,LOCAL_CHARSET);
+
+      if(tag == 1)
+         {
+         m_type = "RFC822";
+         }
+      else if(tag == 2)
+         {
+         m_type = "DNS";
+         }
+      else if(tag == 6)
+         {
+         m_type = "URI";
+         }
+      }
+   else if(tag == 4)
+      {
+      X509_DN dn;
+      std::multimap<std::string, std::string> nam;
+      BER_Decoder dec(obj.value);
+      std::stringstream ss;
+
+      dn.decode_from(dec);
+      ss << dn;
+
+      m_name = ss.str();
+      m_type = "DN";
+      }
+   else if(tag == 7)
+      {
+      if(obj.value.size() == 8)
+         {
+         const std::vector<byte> ip(obj.value.begin(),obj.value.begin() + 4);
+         const std::vector<byte> net(obj.value.begin() + 4,obj.value.end());
+         m_type = "IP";
+         m_name = ipv4_to_string(load_be<u32bit>(ip.data(),0)) + "/" + ipv4_to_string(load_be<u32bit>(net.data(),0));
+         }
```

Comment:
If the value is 32 bytes it is IPv6. I'm not sure if it makes sense to ignore such a field outright as here, or encode it for inspection even if matching on it is not supported. But either way if the field is not 8 or 32 bytes, it should be rejected